assisted-events-scrape/workers/cluster_events_worker.py

```diff
@@ -49,34 +50,47 @@ def store_events_for_cluster(self, cluster: dict) -> None:
             if "hosts" not in cluster or len(cluster["hosts"]) == 0:
                 cluster["hosts"] = self.__get_hosts(cluster["id"])
             events = self.__get_events(cluster["id"])
-            component_versions = self._ai_client.get_versions()
+            component_versions = self.__get_versions()
             self._store_normalized_events(component_versions, cluster, events)
             self.cluster_events_storage.store(component_versions, cluster, events)
             self._config.changes.set_changed()
             log.debug(f'Storing events for cluster {cluster["id"]}')
         except Exception as e:
             self.__handle_unexpected_error(e, f'Error while processing cluster {cluster["id"]}')
 
+    @retry(ApiException, delay=1, tries=3, backoff=2, max_delay=4, jitter=1)
+    def __get_versions(self):
+        def _internal_get_versions():
+            return self._ai_client.get_versions()
+
+        return handle_4XX_apiexception(
+            _internal_get_versions,
+            message_404="Versions not found. This should never happen",
+            default_return_value=[]
+        )
+
+    @retry(ApiException, delay=1, tries=3, backoff=2, max_delay=4, jitter=1)
     def __get_events(self, cluster_id: str):
-        events = []
-        try:
-            events = self._ai_client.get_events(cluster_id, categories=EVENT_CATEGORIES)
-        except ApiException as e:
-            if e.status != 404:
-                raise e
-            log.debug(f'Events for cluster {cluster_id} not found')
-        return events
+        def _internal_get_events():
```

Comment:
@osherdp can you think a cleaner way to do this? The aim is that my wrapper would call this function with its arguments

Reply:
I prefer to not have any internal functions defined inside functions, just for the sake of having an easier way to test them and because it isn't Pythonic to my eyes
other than that, handling looks great

Reply:
yeah, I agree. My question was: "What's the Pythonic way of doing this?" I can't figure out a nice way of doing it. Maybe doing it as a decorator, although it would be most likely more complex code. WDYT?

Reply:
Well, I was trying to implement the decorator, but I've realized one of the issue is the custom error message which would be good to maintain